### PR TITLE
Clean up AnalyzePolicy

### PR DESCRIPTION
**Pull request by dhollman**
Right now, the size of the code for `AnalyzePolicy` scales quadraticly with the number of traits we analyze. With this change, the scaling is linear and adding a new trait doesn't have to touch any other code. There's nothing fancy going on here; we're just using partial specialization to detect the various traits. The one trick that was missing before is that we have to extract the work tag using an unspecialized template that defers to a base with a different name (which is partially specialized) before going back to the analysis of the rest of the traits.

### CODE REVIEW

**Comment by dalg24** [comment]
Retest this please

**Review by masterleinad — COMMENTED** [review]
Otherwise, this looks good AFAICT.

File: core/src/impl/Kokkos_AnalyzePolicy.hpp [hunk 1]
```diff
@@ -52,212 +52,183 @@
...
-template <typename PolicyBase>
-struct SetIsGraphKernel {
-  using type = PolicyTraitsBase<
-      typename PolicyBase::execution_space, typename PolicyBase::schedule_type,
-      typename PolicyBase::work_tag, typename PolicyBase::index_type,
-      typename PolicyBase::iteration_pattern,
-      typename PolicyBase::launch_bounds,
-      typename PolicyBase::work_item_property, std::true_type>;
+//------------------------------------------------------------------------------
+// Work item propoerty case
```
Comment (dalg24) [inline core/src/impl/Kokkos_AnalyzePolicy.hpp:133]:
```suggestion
// Work item property case
```

File: core/src/impl/Kokkos_AnalyzePolicy.hpp [hunk 2]
```diff
@@ -52,212 +52,183 @@
...
+//------------------------------------------------------------------------------
+// Defaults, for the traits that aren't yet handled
 
-  using schedule_type =
-      typename std::conditional<is_void<typename Base::schedule_type>::value,
-                                Schedule<Static>,
-                                typename Base::schedule_type>::type;
+// A tag class for dependent defaults that must be handled by the
+// PolicyTraitsWithDefaults wrapper, since their defaults depend on other traits
+struct dependent_policy_trait_default;
```
Comment (dalg24) [inline core/src/impl/Kokkos_AnalyzePolicy.hpp:189]: Does not need a definition?

**Review by dalg24 — APPROVED** [review]
Fine by me

**Comment by masterleinad** [comment]
Apparently `MSVC` has some problems with this pull request. I tried to generate somewhat better error messages in https://github.com/masterleinad/kokkos/tree/analyze-policy-cleanup (with error messages in https://ci.appveyor.com/project/masterleinad/kokkos-5apfq/builds/36136850). One of them is saying
> [00:07:46] C:\projects\source\core\src\impl/Kokkos_AnalyzePolicy.hpp(172,37): error C2338: Kokkos Error: More than one work tag given [C:\projects\source\build\core\unit_test\KokkosCore_UnitTest_Serial1.vcxproj]
[00:07:46] C:\projects\source\core\src\impl/Kokkos_AnalyzePolicy.hpp(189): message : see reference to class template instantiation 'Kokkos::Impl::TAssertEquality<void,_Ty>' being compiled [C:\projects\source\build\core\unit_test\KokkosCore_UnitTest_Serial1.vcxproj]
[00:07:46]           with
[00:07:46]           [
[00:07:46]               _Ty=Kokkos::Impl::AnalyzePolicy<void,>::launch_bounds
[00:07:46]           ]

so apparently `launch_bounds` becomes a work tag at some point (for MSVC).

**Comment by masterleinad** [comment]
```
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp: In member function 'Kokkos::Impl::AnalyzePolicy<void>& Kokkos::Impl::AnalyzePolicy<void>::operator=(const Kokkos::Impl::PolicyTraitsWithDefaults<Other>&)':
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:329:56: error: no return statement in function returning non-void [-Werror=return-type]
   AnalyzePolicy& operator=(PolicyTraitsWithDefaults<Other> const&) {}
                                                        ^
```

**Pull request merged**